lms/djangoapps/course_api/blocks/transformers/block_completion.py

```diff
@@ -43,9 +45,43 @@ def get_block_completion(cls, block_structure, block_key):
     def collect(cls, block_structure):
         block_structure.request_xblock_fields('completion_mode')
 
+    def mark_complete(self, course_block_completions, latest_completion_block_key, block_key, block_structure):
+        """
+        Helper function to mark a block as 'complete' as dictated by
+        course_block_completions (for problems) or all of a block's children being complete.
+        This also sets the 'resume_block' field as that is connected to the latest completed block.
+
+        :param course_block_completions: dict[course_completion_object] =  completion_value
+        :param latest_completion_block_key: block key for the latest completed block.
+        :param block_key: A opaque_keys.edx.locator.BlockUsageLocator object
+        :param block_structure: A BlockStructureBlockData object
+        """
+        if block_key in course_block_completions:
+            block_structure.override_xblock_field(block_key, self.COMPLETE, True)
+            if block_key == latest_completion_block_key:
+                block_structure.override_xblock_field(block_key, self.RESUME_BLOCK, True)
+
+        children = block_structure.get_children(block_key)
+        if (children and all(block_structure.get_xblock_field(child_key, self.COMPLETE)
```

Comment:
This would probably be more readable if you split it into some smaller generators.

```
non_discussion_children = (child_key in children if block_structure.get_xblock_field(child_key, 'category') != 'discussion')
child_complete = (block_structure.get_xblock_field(child_key, self.COMPLETE) for child_key in non_discussion_children)
if (children and all(child_complete)): ...
```

Reply:
Or, alternately, maybe just use `children = [block_structure[child_key] for child_key in block_structure.get_children(block_key)]` so that you have all the fields, and can dispense with the `block_structure.get_xblock_field` calls in the rest of the function?

Reply:
Oh interesting. Is using `get_xblock_field` not the expectation? That would make a lot of simplifications in the `get_course_assignments` function.

Reply:
In testing this out, it doesn't provide the default value functionality since it turns it into a direct attribute access so I think calling it this way isn't too bad. I'll do the refactor though to try and make it more readable by separating out the steps.